processing/calculations.py

```diff
@@ -0,0 +1,129 @@
+from statistics import mean, stdev
+
+
+def start_calculate_risk(process_dict: dict, metrics_dict: dict) -> dict:
+    """
+    The entry function and handler for the risk calculations. The risk score is calculated and returned with all related
+    values.
+
+    :param metrics_dict: A dict containing all relevant data for the metrics
+    :type metrics_dict: Dict[str: Any]
+    :param process_dict: A dict containing all relevant data of a process to calculate the risk score
+    :type process_dict: Dict[str: Any]
+    :return: Dict[str: Any]
+    """
+
+    current_values = calculate_current_values(process_dict)
+    compared_values = compare_actual_target_metrics(
+        current_values, metrics_dict)
+    full_process_dict = calculate_risk_score(compared_values)
+
+    return full_process_dict
+
+
+def calculate_current_values(process_dict: dict) -> dict:
+    """
+    Function that extracts the current metrics values from a process dict
+
+    :param process_dict: A dict containing all relevant data of a process to calculate the risk score
+    :type process_dict: Dict[str: Any]
+    :return: Dict[str: Any]
+    """
+
+    # get all metric values
+    component_metrics = {}
+    for components in (process_dict["process"]["components"]):
+        for key, value in components["metrics"].items():
+            if key not in component_metrics:
+                component_metrics[key] = [value]
+            else:
+                component_metrics[key].append(value)
+
+    # prepare output dict
+    output_dict = {'success': process_dict["success"],
+                   'process': process_dict["process"],
+                   'actual_target_metrics': {}}
+
+    calculations = output_dict['actual_target_metrics']
+
+    for key, value in component_metrics.items():
+        calculations[key] = {}
+
+        # calculate actual metrics
+        calculations[key]['actual'] = {"total": sum(value),
+                                       "min": min(value),
+                                       "max": max(value),
+                                       "avg": mean(value),
+                                       }
+
+        # more than one component has this metric, thus a standard deviation can be calculated
+        if len(value) > 1:
+            calculations[key]['actual'].update({"std_dev": stdev(value)})
+
+        # only one component has this metric, thus no standard deviation can be calculated
+        else:
+            calculations[key]['actual'].update({"std_dev": None})
+
+        # get amount of components
+        calculations[key]["count_component"] = len(value)
+
+        # calculate and get target metrics
+        calculations[key]['target'] = {'avg': process_dict['target_metrics'][key],
```

Comment:
Hier tritt ein Fehler auf @Nico3001 oder @nicolasschmid0 . Das Problem ist wohl, dass er andere Metriken in dem process_dict["target_metrics"] als in dem component_metrics hat. Dadurch kann er nicht den richtigen key einsetzen und schmeisst einen raus. Bei Rückfragen komm gerne auf mich zu.

Reply:
Haben wir drüber gesprochen: wird nie der Fall sein. Auf Prozessebene müssen alle Metriken eingegeben werden, um gespeichert werden zu können.

Reply:
Das verstehe ich ehrlich gesagt nicht ganz. Auf Prozessebene müssen die Metriken der Komponenten eingegeben werden?

Reply:
Nein, die Prozess-Metriken und da diese immer alle eingegeben werden müssen und hier die Abfrage nach den Metriken der Komponenten auf die Metriken des Prozesses erfolgt, müssten diese immer gefunden werden. Ich kann auch gerne hier morgen nochmal nachschauen :)

Reply:
Also bei mi funktioniert es mit folgendem Code nicht: 
print(core.get_process({"uid": "153fd7899gfe40569c0c35b9b983c7be"}))

Wenn du sagst das ist gewollt oder funktioniert alles, gebe ich gerne frei. Kann das mit meinen Tests aber nicht nachvollziehen.

Reply:
![image](https://user-images.githubusercontent.com/44087726/114146909-6b190a00-9918-11eb-8b44-aca505f5e7ba.png)

Wie du hier siehst (debugging deines Aufrufs) hat schon die erste Komponente Metriken, die der Prozess nicht hat, das wird es in der produktiv-Umgebung nicht geben.

Wenn wir im Debugging-Prozess allerdings dem Prozess die entsprechenden target_metrics doch füllen:
{'comment_quality': 2, 'code_review_frequency': 5, 'development_time': 30, 'number_of_lines_of_source_code_loc': 20, 'encryption': 1337.0, 'number_of_administrators': 1337.0, 'test_scope': 1337.0, 'testability': 1337.0, 'error_during_testing': 1337.0, 'pre_post_release_faults_ratio': 1337.0, 'ratio_of_late_corrections': 1337.0, 'test_automation': 1337.0}

![image](https://user-images.githubusercontent.com/44087726/114151181-25ab0b80-991d-11eb-8af6-d4fe3cf5d084.png)

Dann wird auch die Berechnung korrekt ausgeführt, ein score berechnet und actual_target_metrics gefüllt und zurück geliefert.
![image](https://user-images.githubusercontent.com/44087726/114151397-5c812180-991d-11eb-8518-ed55716114f0.png)

Reply:
Genau das war das Problem. Die Listen der Metriken haben nicht übereingestimmt. Dass das Frontend das nicht zulässt war zum jetzigen Zeitpunkt für mich nicht erkennbar.